src/driver/mio/op.rs

```diff
@@ -0,0 +1,197 @@
+use std::io;
+
+use mio::event::Event;
+
+pub use crate::driver::unix_op::*;
+use crate::{
+    buf::{AsIoSlices, AsIoSlicesMut, IoBuf, IoBufMut},
+    driver::{Decision, OpCode},
+    op::*,
+};
+
+/// Helper macro to execute a system call that returns an `io::Result`.
+macro_rules! syscall {
+    ($fn: ident ( $($arg: expr),* $(,)* ) ) => {{
+        #[allow(unused_unsafe)]
+        let res = unsafe { ::libc::$fn($($arg, )*) };
+        if res == -1 {
+            Err(::std::io::Error::last_os_error())
+        } else {
+            Ok(res as _)
+        }
+    }};
+}
+
+/// Execute a system call, if would block, wait for it to be readable.
+macro_rules! syscall_or_wait_writable {
+    ($fn: ident ( $($arg: expr),* $(,)* ), $fd:expr) => {{
+        match syscall!( $fn ( $($arg, )* )) {
+            Ok(fd) => Ok(Decision::Completed(fd)),
+            Err(e) if e.kind() == io::ErrorKind::WouldBlock => Ok(Decision::wait_writable($fd)),
+            Err(e) => Err(e),
+        }
+    }};
+}
+
+/// Execute a system call, if would block, wait for it to be writable.
+macro_rules! syscall_or_wait_readable {
+    ($fn: ident ( $($arg: expr),* $(,)* ), $fd:expr) => {{
+        match syscall!( $fn ( $($arg, )* )) {
+            Ok(fd) => Ok(Decision::Completed(fd)),
+            Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => Ok(Decision::wait_readable($fd)),
+            Err(e) => Err(e),
+        }
+    }};
+}
+
+impl<T: IoBufMut> OpCode for ReadAt<T> {
+    fn pre_submit(&mut self) -> io::Result<Decision> {
+        Ok(Decision::wait_readable(self.fd))
+    }
+
+    fn on_event(&mut self, event: &Event) -> std::io::Result<usize> {
+        debug_assert!(event.is_readable());
+
+        let (ptr, len) = (self.buffer.as_buf_mut_ptr(), self.buffer.buf_len());
+        syscall!(pread(self.fd, ptr as _, len, self.offset as _))
+    }
+}
+
+impl<T: IoBuf> OpCode for WriteAt<T> {
+    fn pre_submit(&mut self) -> io::Result<Decision> {
+        Ok(Decision::wait_writable(self.fd))
+    }
+
+    fn on_event(&mut self, event: &Event) -> std::io::Result<usize> {
+        debug_assert!(event.is_writable());
+
+        let (ptr, len) = (self.buffer.as_buf_ptr(), self.buffer.buf_len());
+        syscall!(pwrite(self.fd, ptr as _, len, self.offset as _))
+    }
+}
+
+impl OpCode for Sync {
+    fn pre_submit(&mut self) -> io::Result<Decision> {
+        Ok(Decision::Completed(syscall!(fsync(self.fd))?))
+    }
+
+    fn on_event(&mut self, _: &Event) -> std::io::Result<usize> {
+        unreachable!("Sync operation should not be submitted to mio")
```

Comment:
Why? There seems to be `sync_*` in tokio.

Reply:
They spawn

Reply:
Then we could just perform a sync fsync. The Windows one is also a sync flush.

Reply:
Already did in pre_submit

Reply:
The logic behind `mio::OpCode` is that when submitting, `pre_submit` gets executed and determines if the op will be submitted into mio. If it's submitted and indeed receives events, `on_event` gets called. In this case, `Sync` is only a `pre_submit` op and will never be submitted into mio, so `on_event` should be `unreachable.`